icu4c/source/common/mlbe.cpp

```diff
@@ -128,135 +122,169 @@ int32_t MlBreakEngine::divideUpRange(UText *inText, int32_t rangeStart, int32_t
     return correctedNumBreaks;
 }
 
-void MlBreakEngine::evaluateBreakpoint(UChar32* elementList, int32_t index, int32_t &numBreaks,
-                                         UVector32 &boundary, UErrorCode &status) const {
+void MlBreakEngine::evaluateBreakpoint(const UnicodeString& inString, int32_t* indexList,
+                                       int32_t index, int32_t codeUts, int32_t& numBreaks,
+                                       UVector32& boundary, UErrorCode& status) const {
     if (U_FAILURE(status)) {
         return;
     }
-
-    UnicodeString feature;
+    int32_t end = 0;
     int32_t score = fNegativeSum;
 
-    if (elementList[0] != INVALID) {
-        // When the key doesn't exist, Hashtable.geti(key) returns 0  and 2 * 0 = 0.
+    if (indexList[0] != -1) {
+        // UW1
+        end = (indexList[1] != -1) ? indexList[1] : codeUts;
+        // When the key doesn't exist, Hashtable.geti(key) returns 0.
         // So, we can skip to check whether fModel includes key featureList[j] or not.
-        score += (2 * fModel.geti(feature.setTo(u"UW1:", 4).append(elementList[0])));
+        score += fModel[0].geti(inString.tempSubString(indexList[0], end - indexList[0]));
     }
-    if (elementList[1] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"UW2:", 4).append(elementList[1])));
+    if (indexList[1] != -1) {
+        // UW2
+        end = (indexList[2] != -1) ? indexList[2] : codeUts;
+        score += fModel[1].geti(inString.tempSubString(indexList[1], end - indexList[1]));
     }
-    if (elementList[2] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"UW3:", 4).append(elementList[2])));
+    if (indexList[2] != -1) {
+        // UW3
+        end = (indexList[3] != -1) ? indexList[3] : codeUts;
+        score += fModel[2].geti(inString.tempSubString(indexList[2], end - indexList[2]));
     }
-    if (elementList[3] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"UW4:", 4).append(elementList[3])));
+    if (indexList[3] != -1) {
+        // UW4
+        end = (indexList[4] != -1) ? indexList[4] : codeUts;
+        score += fModel[3].geti(inString.tempSubString(indexList[3], end - indexList[3]));
     }
-    if (elementList[4] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"UW5:", 4).append(elementList[4])));
+    if (indexList[4] != -1) {
+        // UW5
+        end = (indexList[5] != -1) ? indexList[5] : codeUts;
+        score += fModel[4].geti(inString.tempSubString(indexList[4], end - indexList[4]));
```

Comment:
after your last PR,  UW1 to UW5 (not sure about UW6, please see could you) could be merged into one for loop
```
for (int32_t j = 0 ; j < 5; j++) {
  if (indexList[j] != -1) {
        // UWj
        end = (indexList[j+1] != -1) ? indexList[j+1] : codeUts;
        // When the key doesn't exist, Hashtable.geti(key) returns 0.
        // So, we can skip to check whether fModel includes key featureList[j] or not.
        score += fModel[1].geti(inString.tempSubString(indexList[j], end - indexList[j]));
    }
}
```

Reply:
Done